### PR TITLE
clean: utilize QSaveFile and improve config/favorite/history file saving

### PR DESCRIPTION
**Pull request by shenlebantongying**
Delete `atomic_rename.cc/.hh` :sweat_smile: 

It exists because `QFile::rename` won't overwrite an existing target file.

Win32 -> `MoveFileEx(W)` & stdio.h -> `rename` will do.

Create tmp file -> rename is to avoid writing failures.

QSaveFile does the same, but without the need to manually write code for creating a temporal file and rename.

### CODE REVIEW

File: src/config.cc [hunk 1]
```diff
@@ -2235,12 +2233,8 @@ void save( Class const & c )
 
   QByteArray result( dd.toByteArray() );
 
-  if ( configFile.write( result ) != result.size() )
+  if ( configFile.write( result ) != result.size() || configFile.commit() )
```
Comment (shenlebantongying) [inline src/config.cc:2236]: new code handled the case of rename failure.

File: src/main.cc [hunk 2]
```diff
@@ -498,7 +497,10 @@ int main( int argc, char ** argv )
         return -1;
 
       QString configFile = Config::getConfigFileName();
-      renameAtomically( configFile, configFile + ".bad" );
+      QFile::rename( configFile,
+                     configFile % QStringLiteral( "." )
+                       % QDateTime::currentDateTime().toString( QStringLiteral( "yyyyMMdd_HHmmss" ) )
```
Comment (shenlebantongying) [inline src/main.cc:502]: Time stamped bad/backup files, instead of overwriting the same one.

File: src/config.cc (hunk 1 above)
Comment (xiaoyifang) [inline src/config.cc:2236]: configFile.commit()  or  ! configFile.commit()

Comment (xiaoyifang) [reply src/config.cc:2236]: if `configFile.write( result ) != result.size()` evaluate to true ,the configFile.commit() will get no chance to run.
from the doc of QSaveFile ,the commit method should be called mandatory

Comment (shenlebantongying) [reply src/config.cc]: `!` 

:)

File: src/main.cc (hunk 2 above)
Comment (xiaoyifang) [reply src/main.cc:503]: Can QSaveFile be used to save the backupfile to `configFile+".bad"`

Comment (shenlebantongying) [reply src/main.cc:503]: no

File: src/config.cc (hunk 1 above)
Comment (shenlebantongying) [reply src/config.cc]: No, commit() is mandatory if we want to copy the tmp file to target.

`configFile.write( result ) != result.size()` hints something wrong, thus no need to commit().

Comment (xiaoyifang) [reply src/config.cc]: image

the commit method should be handling the write failure check .


Can this line 
`if ( configFile.write( result ) != result.size() || !configFile.commit() )`
reduce to 
`if(!configFile.commit())` alone.

Comment (xiaoyifang) [reply src/config.cc]: sounds ok.

Comment (shenlebantongying) [reply src/config.cc]: Actually, I think the `configFile.write( result ) != result.size()` is not needed.

QIODevice internally keeps a errorString, and QSaveFile interanlly have another writeError,

The commit() will check both,

So we don't need to manually check

Comment (xiaoyifang) [reply src/config.cc]: configFile.write( result ) is still needed.

 configFile.write( result );
if(configFile.commit()){

Comment (shenlebantongying) [reply src/config.cc]: Fixed,

The rabbit hole:

QFile's write error check https://github.com/qt/qtbase/blob/ae8031b5e72032f9e2884c18cd72639acfd0d1a4/src/corelib/io/qsavefile.cpp#L307

rename failure check https://github.com/qt/qtbase/blob/ae8031b5e72032f9e2884c18cd72639acfd0d1a4/src/corelib/io/qsavefile.cpp#L315

The write error is done by reimplement QIODevice's writeData https://github.com/qt/qtbase/blob/ae8031b5e72032f9e2884c18cd72639acfd0d1a4/src/corelib/io/qsavefile.cpp#L362-L363

**Pull request merged**